src/controllers/hooks/resource-store/index.tsx

```diff
@@ -16,45 +19,58 @@ type UseResourceHookResponse<
   refresh: () => void;
 };
 
-/**
- * Utility hook for using a resource. Will throw a promise if we are using suspense.
- *
- */
+type UseResourceOptions = {
+  routerContext?: RouterStoreContext;
+};
+
 export const useResource = (
-  resource: RouteResource
-): UseResourceHookResponse[] => {
-  /**
-   * @gasparin - We may need to use a hook to get the router so that components using this hook
-   * will respond to route changes, so something like
-   *
-   * const [router] = useRouterStore()
-   */
-  const { route, match, query, location } = getRouterState();
-  const { type, getKey, maxAge } = resource;
+  resource: RouteResource,
+  options?: UseResourceOptions
+): UseResourceHookResponse => {
   const [, actions] = useResourceActions();
-  const key = getKey({ match, query, route, location }, actions.getContext());
-  const [slice] = useResourceStore({
-    type,
-    key,
-  });
-
-  return [
-    {
-      ...slice,
-      update: useCallback(
-        (getNewData: RouteResourceUpdater) => {
-          actions.updateResourceState(type, key, maxAge, getNewData);
-        },
-        [actions, type, key, maxAge]
-      ),
-      refresh: useCallback(() => {
-        actions.getResourceFromRemote(resource, {
-          route,
-          match,
-          query,
-          location,
-        });
-      }, [actions, resource, route, match, query, location]),
+  const [, { getContext: getRouterContext }] = useRouterStoreStatic();
+
+  // Dynamically generate a router subscriber based on the resource:
+  // makes the component re-render only when key changes instead of
+  // after every route change
+  const useKey = useMemo(
+    () =>
+      createHook<
+        EntireRouterState,
+        AllRouterActions,
+        string,
+        RouterStoreContext | void
+      >(RouterStore, {
+        selector: ({ match, query, route }, keyArg): string =>
+          resource.getKey(
+            keyArg != null ? keyArg : { match, query, route },
+            actions.getContext()
+          ),
+      }),
+    [resource, actions]
+  );
+
+  const key = useKey(options?.routerContext!)[0];
+  const [slice] = useResourceStore({ type: resource.type, key });
+
+  const update = useCallback(
+    (updater: RouteResourceUpdater) => {
+      actions.updateResourceState(resource.type, key, resource.maxAge, updater);
     },
-  ];
+    [resource, key, actions]
+  );
+
+  // we keep route context bound to key, so route context won't refresh
+  // unless key changes. This allows refresh to be called on effect cleanup
+  // or asynchronously, when route context might already have changed
+  const routerContext = useMemo(
```

Comment:
do you think we could get an advantage here by using something like `useMemoOne` which guarantees the value won't be updated until the array value changes rather than leaving this open to being cleaned up by React? Same for the `useCallback`.

Reply:
That's not enough. Take for instance `/projects/:projectKey/board/:boardId`: the match object will be `{ projectKey, boardId }` but if with `getKey: ({ match }) => match.projectKey` we don't care about `boardId` value.
So on `boardId` change we don't want to generate new values for `update` and `refresh` because the "slice" they are bound to is still the same.